lib/internal/modules/esm/import_map.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const { isURL, URL } = require('internal/url');
+const { ObjectEntries, ObjectKeys, SafeMap, ArrayIsArray } = primordials;
+const { codes: { ERR_INVALID_IMPORT_MAP } } = require('internal/errors');
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, this.#baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  resolve(specifier, parentURL = this.baseURL) {
+    // Process scopes
+    for (const { 0: prefix, 1: mapping } of this.scopes) {
+      let mappedSpecifier = mapping.get(specifier);
+      if (parentURL.pathname.startsWith(prefix.pathname) && mappedSpecifier) {
+        if (!isURL(mappedSpecifier)) {
+          mappedSpecifier = new URL(mappedSpecifier, this.baseURL);
+          mapping.set(specifier, mappedSpecifier);
+        }
+        specifier = mappedSpecifier;
+        break;
+      }
+    }
+
+    let spec = specifier;
+    if (isURL(specifier)) {
+      spec = specifier.pathname;
+    }
+    let importMapping = this.imports.get(spec);
+    if (importMapping) {
+      if (!isURL(importMapping)) {
+        importMapping = new URL(importMapping, this.baseURL);
+        this.imports.set(spec, importMapping);
+      }
+      return importMapping;
+    }
+
+    return specifier;
+  }
+}
+
+function processImportMap(importMap, baseURL, raw) {
+  // Validation and normalization
+  if (typeof raw.imports !== 'object' || ArrayIsArray(raw.imports)) {
+    throw new ERR_INVALID_IMPORT_MAP('top level key "imports" is required and must be a plain object');
+  }
+  if (typeof raw.scopes !== 'object' || ArrayIsArray(raw.scopes)) {
+    throw new ERR_INVALID_IMPORT_MAP('top level key "scopes" is required and must be a plain object');
+  }
+
+  // Normalize imports
+  for (const { 0: specifier, 1: mapping } of ObjectEntries(raw.imports)) {
```

Comment:
https://github.com/nodejs/node/blob/340b9a827b7dacbbd974faa0b053773901e4ac88/doc/contributing/primordials.md#L144-L179

```suggestion
  const entries = ObjectEntries(raw.imports);
  for (let i = 0; i < entries.length; i++) {
    const { 0: specifier, 1: mapping } = entries[i];
```